src/persist-client/src/internal/metrics.rs

```diff
@@ -2417,6 +2422,177 @@ impl BlobMemCache {
     }
 }
 
+#[derive(Debug)]
+pub struct SemaphoreMetrics {
+    cfg: PersistConfig,
+    registry: MetricsRegistry,
+    fetch: OnceCell<MetricsSemaphore>,
+}
+
+impl SemaphoreMetrics {
+    fn new(cfg: PersistConfig, registry: MetricsRegistry) -> Self {
+        SemaphoreMetrics {
+            cfg,
+            registry,
+            fetch: OnceCell::new(),
+        }
+    }
+
+    /// We can't easily change the number of permits, and the dyncfgs are all
+    /// set to defaults on process start, so make sure we only initialize the
+    /// semaphore once we've synced dyncfgs at least once.
+    async fn fetch(&self) -> &MetricsSemaphore {
+        if let Some(x) = self.fetch.get() {
+            // Common case of already initialized avoids the cloning below.
+            return x;
+        }
+        let cfg = self.cfg.clone();
+        let registry = self.registry.clone();
+        let init = async move {
+            let total_permits = match cfg.announce_memory_limit {
+                // Non-cc replicas have the old physical flow control mechanism,
+                // so only apply this one on cc replicas.
+                Some(mem) if cfg.is_cc_active => {
+                    // We can't easily adjust the number of permits later, so
+                    // make sure we've synced dyncfg values at least once.
```

Comment:
Depends on your definition of "easy", but it's not a wild amount of code: https://github.com/MaterializeInc/materialize/blob/0fd7267e00ef9bbfb1875a52b88fa16e61311908/src/adapter/src/webhook.rs#L392-L419

Reply:
Didn't know we had an example in the repo, but I did think of that technique. The issue is that tokio semaphore has a fixed upper bound on the number of permits (I _think_), which means if you set the dyncfg from big value (e.g. "off") to small value, and then back to big value, it may not be possible to add enough permits for the second big value.

Reply:
There is definitely...

```rust
 pub(crate) const MAX_PERMITS: usize = std::usize::MAX >> 3;
```

But my understanding is that anything that's an illegal value for `add_permits` is also an illegal value for the initial number of permits, so I think if your example flow breaks it would have broken on the initial big value as well?

(Not critical for this PR, but I do want to make sure I'm understanding right...)

Reply:
Not sure I was completely clear, so let me try again. I could be wrong, but this is my read of the docs:
- dyncfg: MAX_PERMITS -> starts the semaphore with `Semaphore::MAX_PERMITS`
- dyncfg: SMALL -> `semaphore.forget_permits(MAX_PERMITS - SMALL)`. This doesn't reduce the number of permits AFAIK, it just leaks them
- dyncfg: MAX_PERMITS -> `semaphore.add_permits(MAX_PERMITS-SMALL)`. This doesn't work because it already had the maximum number of permits from the first step

Reply:
Ah, okay!

The semaphore doesn't explicitly track a total number of permits... just a count of available permits. Acquiring a permit decrements the count. Dropping the permit increments it again. (So to the extent which there is a total number of permits, it's the available count plus all the live permits.)

So while forgetting doesn't decrement the semaphore per se, that's okay because it was already decremented when those permits were acquired. However, forgetting _does_ ensure that those permits are never added back to the pool. So the forget-then-add gets you back to the original state.

Playground demo of this not exploding: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1a3fd33701bfe1a8105e395ee0e790f6

Reply:
Aha! Okay then indeed my read of the docs was wrong, glad we shook this out! (cc @guswynn who I misinformed about this.)

I might do this one as a fast-follow if that's okay. I've done a good amount of testing of the current code, and this is something I'd want to re-deploy to my staging env and play around with before merging